### ISSUE TITLE
Proxy 'port_443': unable to find required use_backend: '<backend-name>'.

### ISSUE THREAD

**Issue by nagliyvred** [docker/dockercloud-haproxy#35]
Hi everyone,

We have an issue migrating existing haproxy configuration from tutum to dockercloud. 
Haproxy container starts, but haproxy is dead. Same configuration works fine on tutum.
haproxy version is 1.3 (latest as of today). Strange thing is - I have a lot of backends, and certain backends work, whilst the others don't. 
Generated configuration as seen in the log:
```
router-dev-1 | 2016-04-16T05:52:46.698724799Z global
router-dev-1 | 2016-04-16T05:52:46.698737886Z   log XXX local0
router-dev-1 | 2016-04-16T05:52:46.698749146Z   log XXX local1 notice
router-dev-1 | 2016-04-16T05:52:46.698760061Z   log-send-hostname
router-dev-1 | 2016-04-16T05:52:46.698770010Z   maxconn 4096
router-dev-1 | 2016-04-16T05:52:46.698780546Z   pidfile /var/run/haproxy.pid
router-dev-1 | 2016-04-16T05:52:46.698793707Z   user haproxy
router-dev-1 | 2016-04-16T05:52:46.698804313Z   group haproxy
router-dev-1 | 2016-04-16T05:52:46.698815108Z   daemon
router-dev-1 | 2016-04-16T05:52:46.698825562Z   stats socket /var/run/haproxy.stats level admin
router-dev-1 | 2016-04-16T05:52:46.698835366Z   ssl-default-bind-options no-sslv3
router-dev-1 | 2016-04-16T05:52:46.698845842Z   ssl-default-bind-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:AES128-GCM-SHA256:AES128-SHA256:AES128-SHA:AES256-GCM-SHA384:AES256-SHA256:AES256-SHA:DHE-DSS-AES128-SHA:DES-CBC3-SHA
router-dev-1 | 2016-04-16T05:52:46.698859144Z defaults
router-dev-1 | 2016-04-16T05:52:46.698869346Z   balance roundrobin
router-dev-1 | 2016-04-16T05:52:46.698878344Z   log global
router-dev-1 | 2016-04-16T05:52:46.698888321Z   mode http
router-dev-1 | 2016-04-16T05:52:46.698897906Z   option redispatch
router-dev-1 | 2016-04-16T05:52:46.698907266Z   option httplog
router-dev-1 | 2016-04-16T05:52:46.698915560Z   option dontlognull
router-dev-1 | 2016-04-16T05:52:46.698924195Z   option forwardfor
router-dev-1 | 2016-04-16T05:52:46.698933645Z   option http-server-close
router-dev-1 | 2016-04-16T05:52:46.698942673Z   timeout connect 5000
router-dev-1 | 2016-04-16T05:52:46.698951699Z   timeout client 50000
router-dev-1 | 2016-04-16T05:52:46.698959195Z   timeout server 50000
router-dev-1 | 2016-04-16T05:52:46.698966630Z listen stats
router-dev-1 | 2016-04-16T05:52:46.698974385Z   bind :1936
router-dev-1 | 2016-04-16T05:52:46.698981745Z   mode http
router-dev-1 | 2016-04-16T05:52:46.698990231Z   stats enable
router-dev-1 | 2016-04-16T05:52:46.698998046Z   timeout connect 10s
router-dev-1 | 2016-04-16T05:52:46.699005833Z   timeout client 1m
router-dev-1 | 2016-04-16T05:52:46.699022957Z   timeout server 1m
router-dev-1 | 2016-04-16T05:52:46.699030927Z   stats hide-version
router-dev-1 | 2016-04-16T05:52:46.699038424Z   stats realm Haproxy\ Statistics
router-dev-1 | 2016-04-16T05:52:46.699046505Z   stats uri /
router-dev-1 | 2016-04-16T05:52:46.699055239Z   stats auth stats:stats
router-dev-1 | 2016-04-16T05:52:46.699064022Z frontend port_443
router-dev-1 | 2016-04-16T05:52:46.699072294Z   bind :443 ssl crt /certs/
router-dev-1 | 2016-04-16T05:52:46.699080471Z   reqadd X-Forwarded-Proto:\ https
router-dev-1 | 2016-04-16T05:52:46.699088222Z   acl is_websocket hdr(Upgrade) -i WebSocket
router-dev-1 | 2016-04-16T05:52:46.699095686Z   acl host_rule_2 hdr(host) -i overlay-dc.trunkplatform.com.au
router-dev-1 | 2016-04-16T05:52:46.699105747Z   acl host_rule_2_port hdr(host) -i overlay-dc.trunkplatform.com.au:443
router-dev-1 | 2016-04-16T05:52:46.699113179Z   use_backend SERVICE_OVERLAY_PROXY if host_rule_2 or host_rule_2_port
router-dev-1 | 2016-04-16T05:52:46.699120166Z frontend port_80
router-dev-1 | 2016-04-16T05:52:46.699127059Z   bind :80
router-dev-1 | 2016-04-16T05:52:46.699134737Z   acl is_websocket hdr(Upgrade) -i WebSocket
router-dev-1 | 2016-04-16T05:52:46.699141994Z   monitor-uri /haproxy?monitor
router-dev-1 | 2016-04-16T05:52:46.699148756Z   acl host_rule_1 hdr(host) -i overlay-dc.trunkplatform.com.au
router-dev-1 | 2016-04-16T05:52:46.699155801Z   acl host_rule_1_port hdr(host) -i overlay-dc.trunkplatform.com.au:80
router-dev-1 | 2016-04-16T05:52:46.699166400Z   use_backend SERVICE_OVERLAY_PROXY if host_rule_1 or host_rule_1_port
router-dev-1 | 2016-04-16T05:52:46.699399378Z INFO:haproxy:Launching HAProxy
router-dev-1 | 2016-04-16T05:52:46.705940056Z INFO:haproxy:HAProxy has been launched(PID: 24)
router-dev-1 | 2016-04-16T05:52:46.706191127Z INFO:haproxy:===========END===========
router-dev-1 | 2016-04-16T05:52:46.710378027Z [ALERT] 106/055246 (24) : Proxy 'port_443': unable to find required use_backend: 'SERVICE_OVERLAY_PROXY'.
router-dev-1 | 2016-04-16T05:52:46.710721532Z [ALERT] 106/055246 (24) : Proxy 'port_80': unable to find required use_backend: 'SERVICE_OVERLAY_PROXY'.
router-dev-1 | 2016-04-16T05:52:46.712622057Z [ALERT] 106/055246 (24) : Fatal errors found in configuration.
```

As you can see backend section is simply missing. 

This is haproxy stackfile:
```
router-dev:
  image: 'dockercloud/haproxy:1.3'
  deployment_strategy: every_node
  environment:
    - 'DEFAULT_SSL_CERT=XXX'
    - EXTRA_SSL_CERTS=XX_DUMMY_CERT
    - MODE=http
    - MONITOR_PORT=80
    - 'MONITOR_URI=/haproxy?monitor'
    - 'OPTION=redispatch, httplog, dontlognull, forwardfor,http-server-close'
    - 'RSYSLOG_DESTINATION=XXXX'
    - 'XX_DUMMY_CERT=<CERT>'
  links:
    - 'overlay-proxy.ops:overlay-proxy'
  ports:
    - '80:80'
    - '443:443'
    - '1936:1936'
    - '5000:5000'
  privileged: true
  restart: always
  roles:
    - global
  tags:
    - dev
```

And this is a stackfile of a linked service:
```
overlay-proxy:
  image: 'quay.io/trunk/overlay-proxy:latest'
  deployment_strategy: high_availability
  environment:
    - FORCE_SSL=true
    - 'VIRTUAL_HOST=http://overlay-dc.trunkplatform.com.au, https://overlay-dc.trunkplatform.com.au'
  expose:
    - '80'
    - '443'
  restart: always
  tags:
    - dev
```
Any help would be much appreciated, as tutum is shutting down and we need to migrate our production ASAP.
Thanks,
Evgeny

**Comment by nagliyvred** [comment]
I did some debugging and managed to get it working after commenting this line: https://github.com/docker/dockercloud-haproxy/blob/master/haproxy/haproxycfg.py#L312

which is an if condition in generating the config
```
if BackendHelper.check_backend_has_routes(backend):
```

**Comment by tifayuki** [comment]
@nagliyvred 

1. In your stackfile, there is a `.` in the links. I don't think docker cloud allows `.` in a service name. Just curious how did you launch a service with the name of `overlay-proxy.ops`

2. Looks like it happens when your linked service is stopped. As you pointed out, backend is not generated if there is no routes inside it, and this makes the haproxy failed to start. I think it is better remove the if condition to prevent haproxy from failing. I will fix it in the next version ASAP.

Thank you.

**Comment by nagliyvred** [comment]
hi @tifayuki ,
thanks for coming back. It's actually <service_name>.<stack_name> which is a format that we've been using in tutum for many months. (yes, we are linking haproxy to services in other stacks to do blue-green deployment with zero downtime).

**Comment by tifayuki** [comment]
@nagliyvred 
ya, `service.stack` makes sense. I never used it personally xD

**Comment by tifayuki** [comment]
Fixed in `dockercloud/haproxy:1.4`
